datalad_next/itertools/reroute.py

```diff
@@ -0,0 +1,232 @@
+""" Functions that allow to route data around upstream iterator """
+
+from __future__ import annotations
+
+from typing import (
+    Any,
+    Callable,
+    Generator,
+    Iterable,
+)
+
+
+__all__ = ['dont_process', 'join_with_list', 'route_in', 'route_out']
+
+
+dont_process = object()
+
+
+def route_out(iterable: Iterable,
+              data_store: list,
+              splitter: Callable[[Any], tuple[Any, Any | None]],
+              ) -> Generator:
+    """ Route data around the consumer of this iterable
+
+    :func:`route_out` allows its user to:
+
+     1. store data that is received from an iterable,
+     2. determine whether this data should be yielded to a consumer of
+        ``route_out``, by calling :func:`splitter`.
+
+    To determine which data is to be yielded to the consumer and which data
+    should only be stored but not yielded, :func:`route_out` calls
+    :func:`splitter`. :func:`splitter` is called for each element of the input
+    iterable, with the element as sole argument. The function should return a
+    tuple of two elements. The first element is the data that is to be
+    yielded to the consumer. The second element is the data that is to be
+    stored in the list ``data_store``. If the first element of the tuple is
+    ``datalad_next.itertools.dont_process``, no data is yielded to the
+    consumer.
```

Comment:
I wonder if the `dont_process` being specific to the first item is more complex then needed and more limiting then necessary. What about making it `Ignore` and have it be applicable to yield and store independently (item 0 and 1 respectively).

Whenever `Ignore` is found, the corresponding item is ignored. Either not yielded, and/or not stored. I think a use case where all items should be yielded and only some routed around is not far-fetched.

PS: in the docstring `element` is used instead of `item`, which seems to be a more standard term.

Reply:
> I wonder if the `dont_process` being specific to the first item is more complex then needed and more limiting then necessary. What about making it `Ignore` and have it be applicable to yield and store independently (item 0 and 1 respectively).
> 

In order to keep the cardinality of the underlying iterator (so that it can be reproduced at the corresponding `route_in`-call), `route_out` will store information for every item from the underlying iterator. If a user does not need to store anything, he could pass `None` (or `b''` or a similar value that indicates "no data" to him) in element 1 of the result. It seems to me that applying `Ignore` to element 1 would be more restrictive 


> Whenever `Ignore` is found, the corresponding item is ignored. Either not yielded, and/or not stored. I think a use case where all items should be yielded and only some routed around is not far-fetched.

iter_annexworktree with a lot of annexed files and only a few non-annexed files is actually that use case. As mentioned above, the number of ignored and not-ignored elements has no influence on the maximum cardinality of the data store. Each item of the underlying iterable has an entry in the data store. 


> PS: in the docstring `element` is used instead of `item`, which seems to be a more standard term.

I changed the docstring to refer to `items` if we talk about members of an `iterable`, and to `elements` if we talk about members of a `tuple` (which is technically also iterable, but we use it here to pass results). Fixed in a9d56e4d2e7d929d1bf8bb52197690c4d716b99e